WIP: modified convert_forward_solution

joewalter opened the pull request:
As mentioned in #4368 the follwing two ways of converting a forward operator to fixed orientation in mne do not give the same result, which is due to use of patch info in convert_forward_solution (if available):
```
# only use convert_forward_solution
forward_convert = convert_forward_solution(
    forward_ori, surf_ori=True, force_fixed=True, copy=True)
```
```
# use convert_forward_solution and _to_fixed_ori
forward_fixed = convert_forward_solution(
    forward_ori, surf_ori=True, force_fixed=False, copy=True)
forward_fixed = _to_fixed_ori(forward_fixed)
```
This PR modifies `convert_forward_solution` by adding a use_cps parameter to indicate whether cps is to be applied or not, and removes _to_fixed_ori.

Still ToDo:
- [x] replace use of `_to_fixed_ori` with `convert_forward_solution` and remove `_to_fixed_ori`

- [x] add use_cps

- [x] adapt tests using use_cps (which might fail now as the way a fixed fwd is computed is changed in this PR)

This closes #4393, closes #4368

agramfort commented:
go go daniel !

agramfort commented:
so far so good !

joewalter commented:
@Eric89GXL is there a reason for not using cortical patch statistics while generating forwards with fixed orientation?

Eric89GXL commented:
I don't think so. I tried to port Matti's code verbatim, so perhaps he has some insight (might be worth a separate email CC'ing me and Alex to clarify, not sure how closely he watches GitHub).

agramfort commented:
yes ask matti.

it's weird not to use it when available.

Eric89GXL commented on mne/beamformer/tests/test_dics.py:
```diff
@@ -44,7 +44,7 @@ def _get_data(tmin=-0.11, tmax=0.15, read_all_forward=True, compute_csds=True):
         forward_surf_ori = _read_forward_solution_meg(fname_fwd, surf_ori=True)
         forward_fixed = _read_forward_solution_meg(fname_fwd, force_fixed=True,
-                                                   surf_ori=True)
+                                                   use_cps=False)
```
is this necessary for the tests to pass?

FWIW these tests have never been all that robust. @britta-wstnr when you have time can you have a look at these to see if there is some way to make them more reliable without increasing (or while decreasing) test time?

joewalter replied:
with `force_fixed=True` the `surf_ori` parameter is ignored, so I deleted it. `use_cps=False` mimics the old behavior of `convert_forward_solution`, so the tests pass without changing any test.

Eric89GXL replied:
I'm asking: if you use what we think should now be the new default (`use_cps=True`) does it break this test?

joewalter replied:
yes, with use_cps=True, test_dics fails

agramfort replied:
what would it take to fix the test_dics?

if cps is meant to be more accurate I guess forcing the use of it could be seen as a bug fix so no need to add the use_cps argument.

@mshamalainen any thought?

joewalter replied:
it's not only the dics test that would fail.  There are basically 3 failing tests if I allow only use_cps=True: 
- test_dics.py (beamformer module)
- test_inverse.py (minimum norm module)
- test_evoked.py (simulation module)

For example, in test_inverse.py for minimum norm, there is a test, which explicitely requires to generate a fixed forward operator without cps.

Do we want to dismiss the option to generate a fixed fwd without cps?

britta-wstnr replied:
@Eric89GXL sure, I will have a look!

Eric89GXL commented on mne/forward/forward.py:
```diff
@@ -392,6 +392,9 @@ def read_forward_solution(fname, force_fixed=False, surf_ori=False,
         implies surf_ori=True.
+    use_cps : bool
+        Whether to use cortical patch statistics to define normal
+        orientations.
```
I wonder if we should instead deprecated `force_fixed` and `surf_ori` here and tell people to use `convert_forward_solution` afterward. @agramfort is the deduplication worth it?

joewalter replied:
ok for me, `read_forward_solution` just calls `convert_forward_solution` at the end, so we could also take it out of the function

Eric89GXL commented on mne/forward/forward.py:
```diff
@@ -560,6 +564,9 @@ def convert_forward_solution(fwd, surf_ori=False, force_fixed=False,
         Whether to return a new instance or modify in place.
+    use_cps : bool
+        Whether to use cortical patch statistics to define normal
+        orientations.
```
Only used when ...

Eric89GXL commented on mne/inverse_sparse/_gamma_map.py:
```diff
@@ -237,8 +236,8 @@ def gamma_map(evoked, forward, noise_cov, alpha, loose=0.2, depth=0.8,
-        forward = deepcopy(forward)
-        _to_fixed_ori(forward)
+        forward = convert_forward_solution(
+            forward, surf_ori=True, force_fixed=True, copy=True)
```
This is an API change. It probably qualifies as a bugfix, but we need to document it since it could change people's results.

joewalter replied:
with `use_cps=True` it mimics the old behavior of using first `convert_forward_solution` and `_to_fixed_ori` afterwards. So results shouldn't change.

Eric89GXL commented on mne/inverse_sparse/mxne_inverse.py:
```diff
@@ -473,8 +473,8 @@ def tf_mixed_norm(evoked, forward, noise_cov, alpha_space, alpha_time,
-        forward = deepcopy(forward)
-        _to_fixed_ori(forward)
+        forward = convert_forward_solution(
+            forward, surf_ori=True, force_fixed=True, copy=True)
```
same about API change here and above

joewalter replied:
the fwd operators should be equivalent, see also above

Eric89GXL replied:
Ahh I see, I misunderstood the flow

joewalter commented:
depending on whether cortical patch statistics should be used for fixed fwds or not, we should change the default for use_cps.

agramfort commented:
@joewalter to move forward on this without breaking anything I proprose you create a private _convert_forward_solution that has a use_cps param that is set to false in the models for which you would break the tests. Then you set convert_forward_solution that does not expose use_cps param and just calls _convert_forward_solution with use_cps set to True.

clear?

joewalter commented:
Clear! But the problem is that the tests typically use `read_forward_solution`, which calls `convert_forward_solution`. So we would need to have a private `_read_forward_solution` and `_convert_forward_solution` in order to not expose the use_cps parameter in the public functions. Is that ok, too?

Eric89GXL commented:
@agramfort WDYT about deprecating the conversion parameters in `read_forward_solution`? It would simplify this a bit.

I don't mind exposing `use_cps` in `convert_forward_solution` because we had it that way for so long, and it's such a core bit of functionality.

joewalter commented:
I introduced `use_cps` in one of my last commits. It's not in master. The problem why I introduced it was that whether cps was used or not was not consistent. 
- Convertion with `surf_ori=True` and `force_fixed=False` was equivalent to using `use_cps=True` now
- Convertion with `surf_ori=True` and `force_fixed=True` was equivalent to using `use_cps=False` now
- Convertion with `surf_ori=True` and `force_fixed=False` followed by _to_fixed_ori was equivalent to using `use_cps=True` now

Eric89GXL commented:
> I introduced use_cps in one of my last commits. It's not in master.

I'm talking about my [previous comment](https://github.com/mne-tools/mne-python/pull/4372#discussion_r126518962), i.e. `force_fixed` and `surf_ori` which are in `master`. There is really no reason for them to be in the reading function. If we remove them, the only place we need to add `use_cps` is in `convert_forward_solution`, which seems fine to me.

joewalter commented:
Ahh I see, sorry, I misunderstood your comment

agramfort commented:
@Eric89GXL what do you suggest?

- deprecate fixed param in read_forward_solution
- force people to use convert_forward_solution which will use cps if present

is that it?

Eric89GXL commented:
Yes, plus still exposing `use_cps` in case they want the old behavior in `convert_forward_solution`.

agramfort commented:
ok so would we be changing silently existing results? if not I am happy,
otherwise we need to document this clearly.

Eric89GXL commented:
We can make a deprecation cycle to change the default. Then there will be no silent changes.

agramfort commented:
sounds like a plan then !

joewalter commented:
But the default was different for fixed and free fwds. To compute a fixed fwd cps was not used, for free with surf_ori=True cps was used if available. Should I set the default to none for the deprecation cycle to leave everything as it was before (different behavior for fixed and free) and we change it to use_cps=True in the next version?

agramfort commented:
yes please

joewalter commented:
I implemented the functions as suggested and added deprecation and future warnings. I have a failing test on this branch (mne/test/test_dipole.py), but this test also fails on the latest master on my machine. Any ideas?

agramfort commented:
redownload testing data?

can you see why CIs fail?

Eric89GXL commented on examples/forward/plot_forward_sensitivity_maps.py:
```diff
@@ -28,7 +28,8 @@
 # Read the forward solutions with surface orientation
-fwd = mne.read_forward_solution(fwd_fname, surf_ori=True)
+fwd = mne.read_forward_solution(fwd_fname)
+fwd = mne.convert_forward_solution(fwd, surf_ori=True)
```
you need `use_cps=False` to avoid a warning (and for future compat) here, right?

joewalter replied:
For reading a free, surface-oriented forward (`surf_ori=True`, `force_fixed=False`) `use_cps=None` is equivalent to `use_cps=True`, that's why I didn't give `use_cps` explicitely here. The warning should only pop up if `force_fixed=True`, as only there the default will be changed.

Eric89GXL commented on examples/inverse/plot_custom_inverse_solver.py:
```diff
@@ -40,8 +40,8 @@
-forward = mne.read_forward_solution(fwd_fname, surf_ori=True)
-
+forward = mne.read_forward_solution(fwd_fname)
+forward = mne.convert_forward_solution(forward, surf_ori=True)
```
can we use `use_cps=True` here?

joewalter replied:
We could, but there is no change, when we change the default from None to True in this case. Should we add use_cps=True in each of these calls even if the behavior will not change when we change the default?

Eric89GXL replied:
Ahh I see. No, just please make sure that any tests and examples that throw warnings or will change behavior when we change the default have their value set explicitly.

joewalter replied:
What about examples? Shall I use `use_cps=False` if fixed forwards are used to make sure that the results won't change when we change the default. Or is it ok to use `use_cps=True` and potentially change the results in order to use the default in the examples.

Eric89GXL commented on examples/inverse/plot_dics_beamformer.py:
```diff
@@ -53,7 +53,8 @@
 # Read forward operator
-forward = mne.read_forward_solution(fname_fwd, surf_ori=True)
+forward = mne.read_forward_solution(fname_fwd)
+forward = mne.convert_forward_solution(forward, surf_ori=True)
```
Does `use_cps=True` work here (and in other examples)? Would be good to use it if we can. If we can't, let's do `use_cps=False` to avoid warnings.

joewalter replied:
I think that all examples will work with `use_cps=True`, there are some tests which require `use_cps=False` while generating a fixed forward.

Eric89GXL commented on mne/forward/forward.py:
```diff
@@ -569,21 +603,34 @@ def convert_forward_solution(fwd, surf_ori=False, force_fixed=False,
+                 'modified in 0.16. The cps (if available) will then be '
+                 'applied by default for both generating forward operators '
+                 'with fixed orientations and surface-oriented forward '
+                 'operators with free orientations.', FutureWarning)
```
"... Set the value explicitly to avoid this warning."

Eric89GXL commented on examples/visualization/plot_ssp_projs_sensitivity_map.py:
```diff
@@ -23,7 +24,9 @@
 
-fwd = read_forward_solution(fname, surf_ori=True)
+fwd = read_forward_solution(fname)
+fwd = convert_forward_solution(fwd, surf_ori=True)
```
Some or all of these tests will break once we change the default `use_cps` behavior, right? And many of these calls will give `FutureWarning` now?

Please change any that emit warnings to use explicit `use_cps=True` (preferably) or `use_cps=False` as necessary. This 1) avoids warnings in tests and examples for now and 2) future-protects us once we make the default change (and then we can also remove a bunch of `use_cps=True` calls).

joewalter replied:
No, the warning will only pop up if `force_fixed=True` and `use_cps=None`. As the default is `force_fixed=False`, we only need to give `use_cps` explicitely, while generating fixed forwards. I can add that.

joewalter commented:
I removed most of the new warnings in the tests by giving use_cps explicitely. Need to double check that none of the compute functions changes behavior silently.

agramfort commented:
tell me when to look

joewalter commented:
@agramfort can you have a look?

I have one hardcoded `use_cps=False` in `_make_forward.py` in the creation of dipole forwards. In `mxne_inverse` and `_gamma_map`, I hardcoded `use_cps=True` in order keep the old behavior (we werde already applying cps in these algorithms while generating fixed fwd).

There are two tests in `test_forward.py`, which use `read_forward_solution` with surf_ori or force_fixed not None, which are causing warnings. These tests should be removed (in 0.16) when these parameters are deprecated.

Eric89GXL commented:
@joewalter what you describe sounds good to me

agramfort commented on mne/forward/forward.py:
```diff
@@ -535,18 +544,41 @@ def read_forward_solution(fname, force_fixed=False, surf_ori=False,
+            fwd['source_nn'] = np.concatenate([_src['nn'][_src['vertno'], :]
+                                              for _src in fwd['src']], axis=0)
+            fwd['source_ori'] = FIFF.FIFFV_MNE_FIXED_ORI
+            fwd['surf_ori'] = True
```
this block is apparently not tested.

agramfort commented:
besides the coverage issue

please update what's new and let's merge

I'll look at your other PR once this one is merged.

Eric89GXL commented:
@agramfort okay for you to merge this once Travis comes back happy?

agramfort commented:
yes but I wan't a what's new entry for this. This is an important API change.

joewalter commented:
@agramfort @Eric89GXL I think I found a reason why cps was not applied by default for generating fixed fwds. The problem is that if you save a fixed fwd, `fwd['source_nn']` is not saved. While reading, `fwd['source_nn']` is generated from `fwd[‘src’]`. If a fixed fwd is generated with `use_cps=True`,  `source_nn` must not be equivalent to the normal orientations at the source locations stored in `fwd[‘src’]`. So the reading might change the orientation of the dipoles in `fwd['source_nn']`. I think that modifying `fwd[‘src’]` while saving is not an option. Any idea?

agramfort commented:
could we save fwd['source_nn'] ? would this solve the problem?

joewalter commented:
Yes, that should solve the problem. Can we save it without losing compatibility with other MNE tools?

agramfort commented:
we can always default to src['nn'] right?

joewalter commented:
it could be sufficient to add a new FIFF constant `FIFF.FIFFV_MNE_FIXED_ORI_CPS` and save it to `FIFF.FIFF_MNE_SOURCE_ORIENTATION`. While reading, we reapply the patch_info if `fwd['source_ori']==FIFF.FIFFV_MNE_FIXED_ORI_CPS` to generate the the true `source_nn`, otherwise we simply take the orientation from `src['nn']`

joewalter commented:
I implemented the idea mentioned in my last comment.
There are two problems:
- the test in line 165 fails while comparing `source_nn` with atol<=1e-7 although the same code and src['nn'] is used to compute the normal orientation with cps
- compatibility issues (as this FIFF constant would only be known in MNE python)

Eric89GXL commented:
> the test in line 165 fails while comparing source_nn with atol<=1e-7

What tolerance makes it pass? If it's e.g. `1e-6` it's probably okay but if it's `1e-2` it's probably not.

Eric89GXL commented:
> compatibility issues (as this FIFF constant would only be known in MNE python)

We have some procedure for contacting the Elekta folks and adding new constants, so this is doable. @jaeilepp might be able to help as he has done it most recently.

joewalter commented:
> What tolerance makes it pass? If it's e.g. 1e-6 it's probably okay but if it's 1e-2 it's probably not.

I currently use atol=1e-6 and rtol=0.0.

Eric89GXL commented:
Six decimal places seems acceptable for normal vectors

Is this good to go then?

joewalter commented:
No, it still uses an additional new FIFF constants. Maybe we can find a better way to get this working without this hack.

joewalter commented:
The previously failing test passes now with the default atol and rtol. I added an additional test to cover an untested io screnario. 

Thinking about the API of reading and writing fwd, I think it is misleading that the fwd is read from disk is not the same as the one written (e.g. as we rotate back to non local coordinates while writing). Can we do better (e.g. by adding FIFF constants which can be read to automatically convert to the fwd before writing it to disk)? Or at least add a clear warning that this is the case?

Eric89GXL commented:
> The previously failing test passes now with the default atol and rtol.

What did you change to make it pass? Was the logic wrong, or some calculation inaccurate? If the former, it's interesting to know that using the patch info only changed `nn` values at the 6th decimal place.

> I think it is misleading that the fwd is read from disk is not the same as the one written

If the standard is "forward solutions are always stored on disk in X/Y/Z RAS", then there is no ambiguity or misleading going on. To me this seems like a code-documentation issue, not a user-facing one (they shouldn't need to care, right?)

joewalter commented:
> What did you change to make it pass? Was the logic wrong, or some calculation inaccurate? If the former, it's interesting to know that using the patch info only changed nn values at the 6th decimal place.

Actually I found that the test failed when patch info was not available, so when standard nn was used. What I did was that I recomputed the source_nn using SVD. I changed it now to simply copying it from src. And I added an additional test, which uses a fwd with patch info, to check that this also works.

> If the standard is "forward solutions are always stored on disk in X/Y/Z RAS", then there is no ambiguity or misleading going on. To me this seems like a code-documentation issue, not a user-facing one (they shouldn't need to care, right?)

I agree it's about code-documentation. Is the info "forward solutions are always stored on disk in X/Y/Z RAS" shown to the user somewhere? I just saw that the rotation in write_forward_solution is silent.
My concern is that if a user converts a fwd to surf_ori and then saves it, the fwd is no longer surface-oriented.

So with
```
fwd = convert_forward_solution(fwd, surf_ori=True)
write_forward_solution(fname, fwd)
fwd2 = read_forward_solution(fname)
```
fwd and fwd_read are not the same, right?

joewalter commented:
Now the only open issue in this PR is the additional FIFF constant that I introduced to be able to check if cps was used to generate a fixed fwd.

Eric89GXL commented:
> fwd and fwd_read are not the same, right?

Ahh I see what you mean. In this case it is a user-documentation level problem.

I don't think we can reasonable change this behavior because it's pretty invasive (many tools and maybe other packages would need to change in ways bigger than we are currently proposing). So it's better here to document it properly so users know that proper round-trip might require `convert_forward_solution`.

The constant seems okay to me, but I don't think @agramfort has commented on that yet.

agramfort commented:
no objection with FIFF constant

can you detect than a user does something wrong? Please help me with a snippet to understand what's a bad practice from a user perspective.

did not find the time to read all the diff in the PR

joewalter commented:
The FIFF constant is only an issue of compatibility with other MNE products. As long as we are in mne-python there should not be a problem.

@agramfort what we need is a clear description or warning that forward operators read from disk are either with a fixed orientation (and thus surface-oriented) or with a free orientation and then in xyz orientation. So to get a surface-oriented forward, the user needs to apply convert_forward_solution. The fwd is not converted automatically while reading it from disk.

agramfort commented:
and if I write a surface-oriented forward it gets put back in xyz?

if so I don't see how this can be bug prone.

joewalter commented:
> and if I write a surface-oriented forward it gets put back in xyz?

Isn't that the current behavior in master? That's why I said we need to make a clear statement that the loaded forward is always in xyz and convert_forward_solution must be applied. Or we add one more FIFF constant (for free surface-oriented) and automatically convert to this state while reading.

joewalter commented:
Alternatively, we go back to exposing all parameters including use_cps in read_forward_solution (as discussed earlier). That way, the user has to explicitely give the desired configuration for the forward operator after reading. That might be more error prone than hoping that the user is aware of the fact that the fwd is not surface-oriented and that convert_forward_solution must be applied.

Eric89GXL commented:
Let's just make it clear that free-orientation is always XYZ, and we're done right?

Eric89GXL commented:
(when saved to disk and read from disk I mean)

agramfort commented:
got it :) ... I think. To make sure:

if I do:

forward = read_forward_solution(...)
forward_ori = convert_forward_solution(forward, surf_ori=True,
force_fixed=False, copy=True)
write_forward_solution('tmp-fwd.fif', forward_ori)
forward = read_forward_solution('tmp-fwd.fif')

I may think I have a surface oriented forward but I don't as it's revert on
save?
did I get it right?

Eric89GXL commented:
Yep, okay with you as long as it's documented properly?

agramfort commented:
yes and maybe get a warning if you save a surface oriented forward?

Eric89GXL commented:
>  maybe get a warning if you save a surface oriented forward?

+1. This warning will probably / hopefully be rare for people, as most people probably generate (non-surf-ori) then immediately save.

joewalter commented:
I added notes in `read_forward_solution` and `write_forward_solution` as well as a warning in  `write_forward_solution`, to tell the user that the rotation to surface orientation is inverted before saving the forward operator.

@agramfort are the 3 new lines in what's new ok for you, or do you want to have more info.

Eric89GXL commented on doc/whats_new.rst:
```diff
@@ -75,6 +75,12 @@ Changelog
 
     - Add option for embedding SVG instead of PNG in HTML for :class:`mne.Report` by `Eric Larson`_
 
+    - Deprecate force_fixed and surf_ori in :func:`mne.forward.read_forward_solution` by `Daniel Strohmeier`_
```
goes under API not Changelog

Eric89GXL commented on mne/forward/forward.py:
```diff
@@ -410,6 +410,12 @@ def read_forward_solution(fname, force_fixed=None, surf_ori=None,
+    .. note:: Forward solutions with free orientation are always stored on disk
+              in X/Y/Z RAS coordinates. To obtain surface-oriented forward
+              operators after reading the forward solution with
+              read_forward_solution, please apply convert_forward_solution
```
```
:func:`convert_forward_solution`
```
same with `read_forward_solution`

These get rendered as docs online and it's nice if they link properly

Eric89GXL commented on mne/forward/forward.py:
```diff
@@ -778,6 +784,13 @@ def write_forward_solution(fname, fwd, overwrite=False, verbose=None):
     --------
     read_forward_solution
+
+    .. note:: Forward solutions with free orientation are always stored on disk
```
This should be a proper [Notes section](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

Eric89GXL commented:
Otherwise LGTM

joewalter commented:
@agramfort @Eric89GXL I updates whats_new and added proper Notes sections.

Still not sure if adding the FIFF constant is the best way here. However, the alternative that I see is to add `fwd['_orig_source_ori']` while reading a forward (copy `fwd['source_ori']`). Then we could always save `fwd['_orig_sol']` and `fwd['_orig_source_ori']`. So if `fwd['_orig_source_ori']` indicates fixed (e.g. the original forward was generated with MNE-C with the fixed option), we save the fixed solution and we can also reload this fixed later. However if `fwd['_orig_source_ori']` indicates free orientation, then we save the original free orientation forward solution and are free to apply any transformation later with `convert_forward_solution`. Doing it this way, we'll keep compatibility to other MNE tools. However, we'd need make clear that the user is aware of the status of `fwd['source_ori']` after reading.

Eric89GXL commented:
> Still not sure if adding the FIFF constant is the best way here. However, the alternative that I see ...

If there is a way to get the existing desired behaviors by adding a private property `fwd['_whatever']` instead of using a new constant, that's preferred. (We try to add new constants sparingly.) It sounds like you're saying this is possible, which is great.

> However, we'd need make clear that the user is aware of the status of fwd['source_ori'] after reading.

We already have new documentation to talk about this, right? But the difference is that now we'd need to supplement the docs to say that a fixed-ori forward computed not using CPS will be saved in fixed ori (surface RAS), whereas fixed-ori forward computed using CPS will be saved free ori (XYZ)? That's not the cleanest thing, but in practice I don't think it should matter too much because people should hopefully prefer the with-CPS method anyway.

agramfort commented:
>
> If there is a way to get the existing desired behaviors by adding a
> private property fwd['_whatever'] instead of using a new constant, that's
> preferred. (We try to add new constants sparingly.) It sounds like you're
> saying this is possible, which is great.
>

+1

Eric89GXL commented:
@joewalter you'll also need a rebase

larsoner commented:
@joewalter will you have time to give the changes a shot next week or so? It would be great to get this in for release

joewalter commented:
@larsoner I will try to have a look by the end of the week.

larsoner commented:
@joewalter any time for this one?

larsoner commented:
Maybe this could also address #4545 -- we could do it separately but it might create more rebase problems

agramfort commented:
what's left to do here?

@joewalter you need to rebase.

joewalter commented:
To implement my last proposal without adding a new FIFF constant we need to:
- [ ] have fwd['_orig_source_ori']
- [ ] check fwd['_orig_source_ori'] while writing forwards, if it indicates free orientation then save the original forward in XYZ so that we can later apply any tranformation with convert_forward_solution
- [ ] this must also work with cropped forwards (e.g. based on a label), which will for sure need some work
- [ ] modify documentation

larsoner commented:
> this must also work with cropped forwards (e.g. based on a label), which will for sure need some work

By "this" I assume you mean saving? As a first pass we can disallow saving cropped forwards. I doubt many people do it and there shouldn't really be a strong need.

agramfort commented:
ok sounds like a clear plan.

do you have time to give it a try before the release?

it would be really great to have sorted this out before.

joewalter commented:
Here's a first try. Needs more testing, so do not merge. And it's an API change, so I'd be happy for comments whether the documentation is clear enough

Open issues:
- I added `sol_grad` to the functions for restricting forwards. Was there a reason that it was not changed in these functions?
- I modify `_orig_sol` and '_orig_sol_grad' in the functions for restricting forwards by restricting these original solutions, too, which is required for the new way of `write_forward_solution`. Is that change ok?

larsoner commented:
> Was there a reason that it was not changed in these functions?

AFAIK people rarely work with `sol_grad` in MNE-Python so it's likely there is some neglect.

> Is that change ok?

I think you should do the same stuff to `_orig_sol` and `_orig_sol_grad`, which it sounds like is your plan.

joewalter commented:
Basically what I do now is saving `_orig_sol` instead of the `sol` when writing a forward solution. If the `_orig_sol` was generated with fixed orienation, then a fixed orientation fwd is saved. If the `_orig_sol` was generated with free orienation (which should always be the case with forwrds created in MNE-Python), then a free orientation fwd is saved reverting any transformation that might have been appiled.
Functions, which restrict `sol`, now also restrict `_orig_sol` (and `sol_grad` + `_orig_sol_grad`). So after saving such a restricted fwd, which was generated from a free orientation fwd and loading it later with `read_forward_solution`, the restricted forward can still be modified with convert_forward_solution.

larsoner commented:
That sounds right to me

agramfort commented:
fine with me too

agramfort commented on doc/whats_new.rst:
```diff
@@ -97,6 +97,8 @@ Changelog
 
     - Add bootstrapped confidence intervals to :func:`mne.viz.plot_compare_evokeds` (that can for now also be accessed as :func:`mne.stats._boostrap_ci`) by `Jona Sassenhagen`_ and `Denis Engemann`_
 
+    - Remove :func:`mne.forward._to_fixed_ori` to avoid different ways for generating forward solutions with fixed orientation by `Daniel Strohmeier`_
```
no need to document changes in private functions.

larsoner commented:
@joewalter any time to rebase and finish this early next week? It's the last big blocker for release. If not, I can take over and try to finish.

joewalter commented:
Codecov still fails. And I am not sure for restricted forwards.

larsoner commented:
> Codecov still fails.

I think it's okay here.

> I am not sure for restricted forwards.

My suggestion from before is to prevent saving restricted forwards. I doubt many people use it and picking is fast.

larsoner commented:
> My suggestion from before

Actually I don't see it above, maybe I never made this comment :) Looking at it again, from what I can see you've already taken care of these bits in the code. What are you unsure about? Whether or not it's correct? Can we add more tests to check this?

larsoner commented:
@joewalter to save you some pain I'll take over the rebases after #4618 is merged

larsoner commented:
@joewalter I've hit a block in investigating #4618 that I'm 95% sure will be fixed by this. What should we do about the restricted ones?

larsoner commented:
Closing for squashed and rebased #4620
